core/commands/storage/session.go

```diff
@@ -41,42 +49,65 @@ func NewSessionID() (string, error) {
 	return seid.String(), nil
 }
 
-func (s *Session) NewSession(ssID string, fileHash string) error {
+func (s *Session) NewSession(ssID string) {
 	s.Lock()
 	defer s.Unlock()
 
-	ch := make(map[string]*StorageChallenge)
-	session := &Session{
-		Time:      time.Now(),
-		Status:    "init",
-		FileHash:  fileHash,
-		Challenge: ch,
-	}
-	SessionMap[ssID] = session
-	return nil
+	s.Time = time.Now()
+	s.Status = "init"
+	s.ChunkInfo = make(map[string]*Chunk)
+	SessionMap[ssID] = s
+}
+
+func (s *Session) SetFileHash(fileHash string) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.FileHash = fileHash
 }
 
-func (s *Session) SetChallenge(ctx context.Context, n *core.IpfsNode, api coreiface.CoreAPI, cid cid.Cid) (*StorageChallenge, error) {
+func (s *Session) NewChunk(hash string, payerPid peer.ID, recvPid peer.ID, channelID *ledgerPb.ChannelID, price int64) (*Chunk, error) {
 	s.Lock()
 	defer s.Unlock()
 
-	hash := cid.String()
+	chunk, ok := s.ChunkInfo[hash]
+	if !ok {
+		chunk = &Chunk{
+			ChannelID: channelID,
+			Payer:     payerPid,
+			Receiver:  recvPid,
+			Price:     price,
+			State:     "init",
+		}
+		s.ChunkInfo[hash] = chunk
+	}
+	s.Time = time.Now()
+	return chunk, nil
+}
+
+func (c *Chunk) SetChallenge(ctx context.Context, n *core.IpfsNode, api coreiface.CoreAPI, cid cidlib.Cid) (*StorageChallenge, error) {
+	c.Lock()
+	defer c.Unlock()
+
 	var sch *StorageChallenge
 	var err error
 	// if the chunk hasn't been generated challenge before
-	if _, ok := s.Challenge[hash]; !ok {
+	if c.Challenge == nil {
 		sch, err = NewStorageChallenge(ctx, n, api, cid)
 		if err != nil {
 			return nil, err
 		}
-		s.Challenge[hash] = sch
-	} else {
-		sch = s.Challenge[hash]
 	}
-
+	c.Challenge = sch
```

Comment:
Your previous code was correct, not this change...

if `c.Challenge != nil` you are setting it to `nil` here

Please follow the previous if/else pattern